languages/wikipron/src/codes.py

```diff
@@ -0,0 +1,174 @@
+
+    session = requests_html.HTMLSession()
+    language_page = session.get(
+        f"https://en.wiktionary.org/wiki/{lang_title}_language",
+        timeout=10
+    )
+    lang_table = language_page.html.find(
+        '.language-category-info > tbody > tr'
+    )
+
+    i = 0
+    while i < len(lang_table):
+        if "Canonical name" in lang_table[i].text:
+            name = lang_table[i].text
+            name = name.split('\n')[1]
+        elif "Language code" in lang_table[i].text:
+            # Canonical name (should) always be filled already.
+            # We grab Wiktionary language code as our entry point
+            # to compare/confirm with ISO 639 TSV File because
+            # Wiktionary name does not always correspond with ISO language name
+            # Wiktionary: Ancient Greek; ISO: Greek, Ancient (to 1453)
+            code = lang_table[i].text
+            code = code.split('\n')[1]
+            break
+        i += 1
+    return name, code
+
+
+def main():
+    new_languages = {}
+    failed_languages = {}
+    prev_languages_file = open("languages.json", "r")
+    prev_languages = json.load(prev_languages_file)
+    language_codes_file = open("iso-639-3_20190408.tsv", "r")
+    iso_list = csv.reader(language_codes_file, delimiter="\t")
+
+    for lang_page_title, total_pages in cat_members():
+        lang = {}
+        iso639_name = None
+        iso639_code = None
+
+        # lang_page_title may come out with a space as in:
+        # "Category:Ancient Greek"
+        # space replaced with _
+        wiktionary_name, wiktionary_code = scrape_wiktionary_info(
+            lang_page_title.replace(" ", "_")
+        )
+
+        for row in iso_list:
+            # Catches ISO 639-3 (row[0]), ISO-639-2 B (row[1])
+            # and ISO 639-2 T (row[2]), ISO 639-1 (row[3])
+            # Converts all to ISO-639-2 B (row[1]) if available
+            # If not available converts to ISO 639-3 (row[0])
+            if wiktionary_code in row[0:4]:
+                iso639_name = row[6]
+                iso639_code = row[1] if row[1] else row[0]
+                break
+
+        if iso639_code:
+            lang[iso639_code] = {
+                "iso639_name": iso639_name,
+                "wiktionary_name": wiktionary_name,
+                "wiktionary_code": wiktionary_code,
+                "casefold": (
+                    prev_languages[iso639_code]["casefold"]
+                    if iso639_code in prev_languages
+                    else None
+                ),
+                "total_pages": total_pages
+            }
+            new_languages.update(lang)
+        else:
+            lang[wiktionary_code] = {
+                "wiktionary_name": wiktionary_name,
+                "total_pages": total_pages
+            }
+            failed_languages.update(lang)
+        # Reset to beginning of tsv file
+        # Else we will continue for loop from index of last match
+        language_codes_file.seek(0)
```

Comment:
It seems a bit odd to do this. It almost looks like we'd want to read in the language code TSV file once and keep a lookup table in memory, so as to avoid accessing the file in every iteration inside this `foo` loop. But I understand rewriting this would be a bit too substantial, probably not worth for this PR.

Reply:
I agree it is quite inefficient. I'll attempt to implement a faster solution along the lines of your suggestion at some point in the future.